File: java/src/test/java/com/leecampbell/cod/domain/DisbursingFundsTest.java
package com.leecampbell.cod.domain;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.temporal.ChronoUnit;
import java.util.UUID;

import com.leecampbell.cod.domain.contracts.*;
import com.leecampbell.cod.domain.model.Loan;

import org.junit.Test;

public final class DisbursingFundsTest {

    private final OffsetDateTime _createdDate = OffsetDateTime.of(2017, 06, 05, 04, 30, 20, 0, ZoneOffset.ofHours(10));
    private final BigDecimal _loanAmount = BigDecimal.valueOf(2000);
    private final OffsetDateTime _transactionDate;
    private final DisburseLoanFundsCommand cmd;
    private final Loan loan;

    public DisbursingFundsTest() {
        
        loan = new Loan(UUID.randomUUID());
        loan.ApplyEvent(new LoanCreatedEvent(_createdDate, new Duration(12, DurationUnit.Month), PaymentPlan.Weekly, _loanAmount));
        loan.ApplyEvent(new LoanCustomerContactChangedEvent("bob", "0444444444", "0812341234", "10 Random Street"));
        loan.ApplyEvent(new LoanBankAccountChangedEvent("066-000", "12345678"));


        _transactionDate = _createdDate.plus(1, ChronoUnit.HOURS);
        cmd = new DisburseLoanFundsCommand(UUID.randomUUID(), loan.id(), _transactionDate);
        loan.DisburseFunds(cmd);
    }

    @Test
    public void loanDisbursedFundsEventRaised() {
        LoanDisbursedFundsEvent actual = (LoanDisbursedFundsEvent)loan.GetUncommittedEvents()[0];
        assertEquals(_transactionDate, actual.transactionDate());
        assertEquals(_loanAmount.negate(), actual.amount());
    }

    @Test
    public void subsequenceAttemptsToDisburseThrow() {
        DisburseLoanFundsCommand cmd2 = new DisburseLoanFundsCommand(UUID.randomUUID(), loan.id(), _transactionDate);
        try {
            loan.DisburseFunds(cmd2);
            fail("Should reject duplicate calls to loan.DisburseFunds(..)");
        } catch (UnsupportedOperationException e) {
            assertEquals("Funds are already disbursed.", e.getMessage());
        }
    }
}

File: typescript/packages/domain-tests/src/disbursing-funds.test.ts
import { describe, it, before } from "node:test";
import assert from "node:assert/strict";
import {
  BankAccount,
  DisburseLoanFundsCommand,
  DisburseLoanFundsCommandHandler,
  Duration,
  DurationUnit,
  type Event,
  type Handler,
  Loan,
  LoanBankAccountChangedEvent,
  LoanCreatedEvent,
  LoanCustomerContactChangedEvent,
  LoanDisbursedFundsEvent,
  FundsAlreadyDisbursedError,
  PaymentPlan,
  Receipt,
} from "@cod/domain";
import { Specification } from "./specification.js";

const CREATED_DATE = new Date("2017-06-04T18:30:20.000Z");
const LOAN_AMOUNT = 2000;

class DisbursingFunds extends Specification<
  Loan,
  DisburseLoanFundsCommand,
  Receipt
> {
  transactionDate!: Date;

  protected createAggregate(id: string): Loan {
    return new Loan(id);
  }

  protected given(): Event[] {
    return [
      new LoanCreatedEvent(
        CREATED_DATE,
        LOAN_AMOUNT,
        Duration.create(12, DurationUnit.Month),
        PaymentPlan.Weekly,
      ),
      new LoanCustomerContactChangedEvent(
        CustomerContact_create(),
      ),
      new LoanBankAccountChangedEvent(BankAccount.create("066-000", "12345678")),
    ];
  }

  protected when(): DisburseLoanFundsCommand {
    this.transactionDate = new Date(CREATED_DATE.getTime() + 3600_000);
    return DisburseLoanFundsCommand.create({
      commandId: crypto.randomUUID(),
      aggregateId: this.sut.id,
      transactionDate: this.transactionDate,
    });
  }

  protected createHandler(): Handler<DisburseLoanFundsCommand, Receipt> {
    return new DisburseLoanFundsCommandHandler(this.repository);
  }
}

class DisbursingFundsMultipleTimes extends Specification<
  Loan,
  DisburseLoanFundsCommand,
  Receipt
> {
  protected createAggregate(id: string): Loan {
    return new Loan(id);
  }

  protected given(): Event[] {
    return [
      new LoanCreatedEvent(
        CREATED_DATE,
        LOAN_AMOUNT,
        Duration.create(12, DurationUnit.Month),
        PaymentPlan.Weekly,
      ),
      new LoanCustomerContactChangedEvent(
        CustomerContact_create(),
      ),
      new LoanBankAccountChangedEvent(BankAccount.create("066-000", "12345678")),
      new LoanDisbursedFundsEvent(
        new Date(CREATED_DATE.getTime() + 3600_000),
        LOAN_AMOUNT,
        BankAccount.create("066-000", "12345678"),
      ),
    ];
  }

  protected when(): DisburseLoanFundsCommand {
    return DisburseLoanFundsCommand.create({
      commandId: crypto.randomUUID(),
      aggregateId: this.sut.id,
      transactionDate: new Date(CREATED_DATE.getTime() + 3600_000),
    });
  }

  protected createHandler(): Handler<DisburseLoanFundsCommand, Receipt> {
    return new DisburseLoanFundsCommandHandler(this.repository);
  }
}

// Helper to avoid importing CustomerContact in every test
import { CustomerContact } from "@cod/domain";
function CustomerContact_create() {
  return CustomerContact.create(
    "bob",
    "0444444444",
    "0812341234",
    "10 Random Street",
  );
}

describe("Disbursing funds", () => {
  const spec = new DisbursingFunds();
  before(async () => spec.execute());

  it("raises LoanDisbursedFundsEvent", () => {
    assert.strictEqual(spec.produced.length, 1);
    const actual = spec.produced[0] as LoanDisbursedFundsEvent;
    assert.deepStrictEqual(actual.transactionDate, spec.transactionDate);
    assert.strictEqual(actual.amount, -LOAN_AMOUNT);
    assert.strictEqual(actual.disbursedTo.bsb, "066-000");
    assert.strictEqual(actual.disbursedTo.accountNumber, "12345678");
  });
});

describe("Disbursing funds multiple times", () => {
  const spec = new DisbursingFundsMultipleTimes();
  before(async () => spec.execute());

  it("throws FundsAlreadyDisbursedError", () => {
    assert.ok(spec.caught instanceof FundsAlreadyDisbursedError);
    assert.strictEqual(spec.caught!.message, "Funds are already disbursed.");
  });
});
